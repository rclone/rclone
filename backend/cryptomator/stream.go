package cryptomator

import (
	"crypto/rand"
	"errors"
	"fmt"
	"io"
)

const (
	// ChunkPayloadSize is the size of the encrypted payload of each file chunk.
	ChunkPayloadSize = 32 * 1024
)

const (
	lastChunk    = true
	notLastChunk = false
)

// Reader decrypts a Cryptomator file as it is read from.
type Reader struct {
	cryptor contentCryptor
	header  FileHeader
	src     io.Reader

	unread []byte
	buf    []byte

	chunkNr uint64

	err error
}

// NewContentReader creates a new Reader for the file content using the previously file header.
func (c *Cryptor) NewContentReader(src io.Reader, header FileHeader) (*Reader, error) {
	cryptor, err := c.newContentCryptor(header.ContentKey)
	if err != nil {
		return nil, err
	}
	return &Reader{
		cryptor: cryptor,
		header:  header,
		src:     src,
		buf:     make([]byte, EncryptedChunkSize(c, ChunkPayloadSize)),
	}, nil
}

// NewReader reads the file header and returns a Reader for the content.
func (c *Cryptor) NewReader(src io.Reader) (r *Reader, err error) {
	header, err := c.UnmarshalHeader(src)
	if err != nil {
		return
	}
	return c.NewContentReader(src, header)
}

func (r *Reader) Read(p []byte) (int, error) {
	if len(r.unread) > 0 {
		n := copy(p, r.unread)
		r.unread = r.unread[n:]
		return n, nil
	}

	if r.err != nil {
		return 0, r.err
	}
	if len(p) == 0 {
		return 0, nil
	}

	last, err := r.readChunk()
	if err != nil {
		r.err = err
		return 0, err
	}

	n := copy(p, r.unread)
	r.unread = r.unread[n:]

	if last {
		if _, err := r.src.Read(make([]byte, 1)); err == nil {
			r.err = errors.New("trailing data after end of encrypted file")
		} else if err != io.EOF {
			r.err = fmt.Errorf("non-EOF error reading after end of encrypted file: %w", err)
		} else {
			r.err = io.EOF
		}
	}

	return n, nil
}

func (r *Reader) readChunk() (last bool, err error) {
	if len(r.unread) != 0 {
		panic("stream: internal error: readChunk called with dirty buffer")
	}

	in := r.buf
	n, err := io.ReadFull(r.src, in)

	switch {
	case err == io.EOF:
		// TODO
		// return false, io.ErrUnexpectedEOF
		return true, nil
	case err == io.ErrUnexpectedEOF:
		last = true
		in = in[:n]
	case err != nil:
		return false, err
	}

	ad := r.cryptor.fileAssociatedData(r.header.Nonce, r.chunkNr)
	payload, err := r.cryptor.DecryptChunk(in, ad)
	if err != nil {
		return
	}

	r.chunkNr++
	r.unread = r.buf[:copy(r.buf, payload)]
	return last, nil
}

// Writer encrypts a Cryptomator file as it is written to.
type Writer struct {
	cryptor contentCryptor
	header  FileHeader

	dst       io.Writer
	unwritten []byte
	buf       []byte

	err error

	chunkNr uint64
}

// NewContentWriter creates a new Writer for the file content using the already written file header.
func (c *Cryptor) NewContentWriter(dst io.Writer, header FileHeader) (*Writer, error) {
	cryptor, err := c.newContentCryptor(header.ContentKey)
	if err != nil {
		return nil, err
	}
	w := &Writer{
		cryptor: cryptor,
		header:  header,
		dst:     dst,
		buf:     make([]byte, EncryptedChunkSize(c, ChunkPayloadSize)),
	}

	w.unwritten = w.buf[:0]
	return w, nil
}

// NewWriter creates and writes a random file header and returns a writer for the file content.
func (c *Cryptor) NewWriter(dst io.Writer) (w *Writer, err error) {
	header, err := c.NewHeader()
	if err != nil {
		return
	}
	err = c.MarshalHeader(dst, header)
	if err != nil {
		return
	}
	return c.NewContentWriter(dst, header)
}

func (w *Writer) Write(p []byte) (n int, err error) {
	if w.err != nil {
		return 0, w.err
	}
	if len(p) == 0 {
		return 0, nil
	}

	total := len(p)
	for len(p) > 0 {
		freeBuf := w.buf[len(w.unwritten):ChunkPayloadSize]
		n := copy(freeBuf, p)
		p = p[n:]
		w.unwritten = w.unwritten[:len(w.unwritten)+n]

		if len(w.unwritten) == ChunkPayloadSize && len(p) > 0 {
			if err := w.flushChunk(notLastChunk); err != nil {
				w.err = err
				return 0, err
			}
		}
	}
	return total, nil
}

// Close flushes the last chunk. It doesn't close the underlying Writer.
func (w *Writer) Close() error {
	if w.err != nil {
		return w.err
	}

	w.err = w.flushChunk(lastChunk)
	if w.err != nil {
		return w.err
	}

	w.err = errors.New("stream.Writer is already closed")
	return nil
}

func (w *Writer) flushChunk(last bool) error {
	if !last && len(w.unwritten) != ChunkPayloadSize {
		panic("stream: internal error: flush called with partial chunk")
	}

	if len(w.unwritten) == 0 {
		return nil
	}

	nonce := make([]byte, w.cryptor.NonceSize())
	_, err := rand.Read(nonce)
	if err != nil {
		return fmt.Errorf("stream: generating nonce failed: %w", err)
	}
	ad := w.cryptor.fileAssociatedData(w.header.Nonce, w.chunkNr)
	out := w.cryptor.EncryptChunk(w.unwritten, nonce, ad)

	_, err = w.dst.Write(out)

	w.unwritten = w.buf[:0]
	w.chunkNr++
	return err
}
