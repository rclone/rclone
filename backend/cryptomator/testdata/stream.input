{
    "example": {
        "CipherCombo": "SIV_CTRMAC",
        "ContentKey": "NSIfk1Uz4Xi0w8P+ulukdoifG5mtFL3TE7zhh/EWZU0=",
        "Nonce": "a/vkY30qi89n93yMT4SQxQ==",
        "MacKey": "lwdzbaZqIfibYj8LSjcUAqpOJfdku/qN13f2OZWpxDk=",
        "Ciphertext": "eAGXNbHreJt00KFj3eVuj/E7snF86cgZ3bCMJOYOdIPRozg3GoQ20t1Fls3KAAEwqNzKpefVO790yZEKKTaTLEy0ZzcO0c7G2i9I2jL3WRM/nE1kZO4YiwEimKoiYnzDVCOqb6DHuGcvp+Zphx/LhrnON/U8fQjijPkU2cr8mU5+3ZMhnsNutHtUygUpK50Y4RS50JJjVlCBzLpD9X2QsnRBODBZybDQ5ZMHYhx/4ogqz0Ww1VZb7n/E9rxZtbbxnMFbcH7TT8cA76mYVxUP+wf57rGH08hLf9J7Ue904w9g/OhNPSeIZbzajafysxpbxTtnsMoOJmRDtbaAiSC4ppHT5fPXzZRPTQLHqhDPtAsJQ2HAhhEhSnp7Bg5bZPA/Vsi2KvCJfGYi1RY9oqMzBYot2fyAc/nlbueGIrM3FokolcJ1QUJsts/Qq2/hMW7gv3zuncsPCssAnXMd0vIhOP8RYmeRx4pStPqvkg51kPAOkDL4xoVA2aQXjZGVXTZ38kkpGFU+okDGRCH4e96w5M2trROq7AvhlxxN/V23/dht7BL9iYcb01XWO7jNt0eh2cXNFunKnq17HsVTRXJOXrawDnFL1fzHT3aSo2T2DQt4SOQS1BMjcZvsD/Fh2JckqbLuPRBwgKiRZYNl0fidwQ9BMMOTxKLhsPv0sTe7JTiqQ5Xjax+eNX+9nwuXO0hXNcg4fqYMLZl4oYRGblufNbpv2kTQMhgzzGBSIyoALHOT0ft68dppRxAx2+tLXaaBJ1o5WCtrLu++KoOSlb106+cZohLBwMFBE1j/93A0wlY4WP4GqrnIG/nvGOs5fPKVulp++Ozl0o/CyLdr9XZboQ=="
    },
    "helloWorld": {
        "CipherCombo": "SIV_CTRMAC",
        "Nonce": "mDCOS8MT0etiyDKaXv8X0Q==",
        "ContentKey": "RgbuFMgbZP051gzgXp8WbQNJ5y50eNyyBL/6qdQNn6c=",
        "MacKey": "lwdzbaZqIfibYj8LSjcUAqpOJfdku/qN13f2OZWpxDk=",
        "Ciphertext": "ug/ko337a8QOjGML9gPpIb6JWlUV5172MFzWWdAANNCh7xiiboSSuv1uU+MGG+bNxXET1fYvEfItU11v"
    }
}
