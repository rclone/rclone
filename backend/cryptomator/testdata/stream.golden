{
    "example": "TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNldGV0dXIgc2FkaXBzY2luZyBlbGl0ciwgc2VkIGRpYW0gbm9udW15IGVpcm1vZCB0ZW1wb3IgaW52aWR1bnQgdXQgbGFib3JlIGV0IGRvbG9yZSBtYWduYSBhbGlxdXlhbSBlcmF0LCBzZWQgZGlhbSB2b2x1cHR1YS4gQXQgdmVybyBlb3MgZXQgYWNjdXNhbSBldCBqdXN0byBkdW8gZG9sb3JlcyBldCBlYSByZWJ1bS4gU3RldCBjbGl0YSBrYXNkIGd1YmVyZ3Jlbiwgbm8gc2VhIHRha2ltYXRhIHNhbmN0dXMgZXN0IExvcmVtIGlwc3VtIGRvbG9yIHNpdCBhbWV0LiBMb3JlbSBpcHN1bSBkb2xvciBzaXQgYW1ldCwgY29uc2V0ZXR1ciBzYWRpcHNjaW5nIGVsaXRyLCBzZWQgZGlhbSBub251bXkgZWlybW9kIHRlbXBvciBpbnZpZHVudCB1dCBsYWJvcmUgZXQgZG9sb3JlIG1hZ25hIGFsaXF1eWFtIGVyYXQsIHNlZCBkaWFtIHZvbHVwdHVhLiBBdCB2ZXJvIGVvcyBldCBhY2N1c2FtIGV0IGp1c3RvIGR1byBkb2xvcmVzIGV0IGVhIHJlYnVtLiBTdGV0IGNsaXRhIGthc2QgZ3ViZXJncmVuLCBubyBzZWEgdGFraW1hdGEgc2FuY3R1cyBlc3QgTG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQuCg==",
    "helloWorld": "aGVsbG8gd29ybGQK"
}
